RFC: A General Recipe for Generic Rules and Natural Tangents (hopefully...) (#449)
JuliaDiff/ChainRulesCore.jl: notes.md

    # Do the primal computation.
    C = A * B

    # "natural pullback": write intuitive pullback, closing over stuff in the usual manner.
    function natural_pullback_for_mul(C̄_natural)
        Ā_natural = C̄_natural * B'
        B̄_natural = A' * C̄_natural
        return NoTangent(), Ā_natural, B̄_natural
    end

    # Make a call to utility functionality which transforms cotangents of C, A, and B.
    # Rule-writing without this utility has similar requirements to `ProjectTo`.
    return C, wrap_natural_pullback(config, natural_pullback_for_mul, C, A, B)
end
```
I'm proposing to coin the term "natural pullback" for pullbacks written within this system, as they're rules written involving natural (co)tangents.

Authors will have to implement two functions for their `AbstractArray` type `P`:
```julia
pullback_of_destructure(::P)
pullback_of_restructure(::P)
```
which are the pullbacks of two functions, `destructure` and `(::Restructure)`, that we'll define later.

The proposed candidates for natural (co)tangents are obtained as follows:
1. natural tangents are obtained from structural tangents via the pushforward of `destructure`,

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Author: >Is that intended? My impression was that we wanted to go from Tangent{Diagonal} to a Diagonal?

I think it's fine to go from `Tangent{Diagonal}` to `Diagonal`.

Two facts:
1. the output of the pullback of restructure must be diagonal,
2. any `AbstractArray` is an accept (co)tangent for an `Array`.

Combining them, we get that it's fine to optimise and return a `Diagonal` since (under the usual `getindex` definition of equality of arrays) it's equal to a diagonal `Array`.

> Interesting, I checked with Zygote and it works (unlike my expectation above) - I think it only works because we have a rule for diag

Agreed -- we would need to manually optimise the pullback for restructure to ensure that this structure is exploited.

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Author: Could you expand on why it's incorrect?

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Reviewer: Ah, sorry - indeed it will always be in the right subspace of Array since its create from a Tangent. So it won't be incorrect, just inefficient if we are not careful.

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Author: Excellent, we have consensus! Is there anything you would suggest we change about the text to clarify?

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Reviewer: should we add something about efficiency? I think not, because it is too much detail. But 🤷

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Author: How about simply
```suggestion
2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`, which can often be manually optimised to utilise structure.
```
?

> 2. natural cotangents are obtained from structural cotangents via the pullback of `(::Restructure)`.

Reviewer: I'd prefer it as is now. We know what the sentence means because of the thread, but it might be confusing to first time reader

Our current `ProjectTo` functionality is roughly the same as the pullback of `destructure`.

In the proposed system, natural (co)tangents remain confined to `rrule`s, and rule authors can choose to work with either natural, structural, or a mixture of (co)tangents.

Other than the headlines at the top, additional benefits of (a correct implementation of) this approach include:
1. no chance of trying to sum tangents failing because one is natural and the other structural,
1. no risk of obstructing AD,
1. rand_tangent can just use structural tangents, simplifying its implementation and improving its reliability,
1. we can probably finally make `to_vec` treat things structurally (although we also need to extend it in other ways), which will also deal with reliability / simplicity of implementation problems,

> 1. generic constructor for composite types easy to implement,

Author: Precisely. Does the following make things clearer:
```suggestion
1. generic constructor for composite types easy to implement since we can be sure of obtaining a structural tangent, and rid ourselves of the "need to defin an adjoint for constructor..." errors we see in Zygote,
```

    _, restructure_C_pb = rrule_via_ad(config, Restructure(C), C_dense)

    function my_mul_generic_pullback(C̄)

        # Recover natural cotangent.
        _, C̄_nat = restructure_C_pb(C̄)

        # Compute pullback using natural cotangent of C.
        Ā_nat = C̄_nat * B_dense'
        B̄_nat = A_dense' * C̄_nat

        # Transform natural cotangents w.r.t. A and B into structural (if non-primitive).
        _, Ā = destructure_A_pb(Ā_nat)
        _, B̄ = destructure_B_pb(B̄_nat)
        return NoTangent(), Ā, B̄
    end

    # The output that we want is `C`, not `C_dense`, so return `C`.
    return C, my_mul_generic_pullback
end
```
All I've done is write out the rrule for differentiating through the equivalent function by hand.
We'll optimise this implementation shortly to avoid e.g. having to densify primals, and computing the same function twice.
`my_mul` in `examples.jl` verifies the correctness of the above implementation.

`destructure` is quite straightforward to define -- essentially equivalent to `collect`. I'm confident that this is always going to be simple to define, because `collect` is always easy to define.

> `Restructure(C)(C_dense)` is a bit trickier. It's the function which takes an `Array` `C_dense` and transforms it into `C`. This feels like a slightly odd thing to do, since we already have `C`, but it's necessary to already know what `C` is in order to construct this function in general -- for example, over-parametrised matrices require this (see the `ScaledMatrix` example in the tests / examples). I'm _reasonably_ confident that this is always going to be possible to define, but I might have missed something.

Author: > I don't know how to do this for e.g. Woodbury, but I think you mentioned you thought about this and know how to do it?

The more I think about it, the less convinced I am that I know how to do this. I think this section on `Restructure` needs re-working / updating to reflect the increased level of uncertainty I have about how easy this is to do. For example, the `ScaledMatrix` example I mention is less straightforward than I originally thought.

> Similarly, two different structured matrices might map to the same dense matrix. In that case I suppose it doesn't matter which structured matrix we get for the correctness?

Yeah, I think that's totally fine. But is this a `destructure` thing rather than `Restructure`?

        # Recover natural cotangent.
        C̄_nat = restructure_C_pb(C̄)

        # Compute pullback using natural cotangent of C.
        Ā_nat = C̄_nat * B'
        B̄_nat = A' * C̄_nat

        # Transform natural cotangents w.r.t. A and B into structural (if non-primitive).
        Ā = destructure_A_pb(Ā_nat)
        B̄ = destructure_B_pb(B̄_nat)
        return NoTangent(), Ā, B̄
    end

    return C, my_mul_generic_pullback
end
```
A few observations:
1. All dense primals are gone. In the pullback, they only appeared in places where they can be safely replaced with the primals themselves because they're doing array-like things. `C_dense` appeared in the construction of `restructure_C_pb`, however, we were using a sub-optimal implementation of that function. Much of the time, `restructure_of_pb` doesn't require `C_dense` in order to know what the pullback would look like and, if it does, it can be obtained from `C`.
2. All direct calls to `rrule_via_ad` have been replaced with calls to functions which are defined to return the things we actually need (the pullbacks). These typically have efficient (and easy to write) implementations.

> 3. `C̄_nat` could be any old `AbstractArray`, because it's conceptually the cotangent for an `Array`. For example, `pullback_of_restructure` for a `Diagonal` returns a `Diagonal`. This is good -- it means we might occassionally get faster computations in the pullback. 

Author: Agreed. I didn't phrase this ideally.

I'm trying to say that it's okay for the `pullback_of_restructure` to return a `Diagonal` in this context. This won't happen if we just use AD though -- we would need to manually arrange for this to happen.

When I wrote this, I think I had already implemented something to make sure that this happens.

> 3. `C̄_nat` could be any old `AbstractArray`, because it's conceptually the cotangent for an `Array`. For example, `pullback_of_restructure` for a `Diagonal` returns a `Diagonal`. This is good -- it means we might occassionally get faster computations in the pullback. 

Author: How about
```suggestion
3. It is permissible for `C̄_nat` to be any old `AbstractArray`, because it's conceptually the cotangent for an `Array`. For example, it is permissible for us to manually optimise `pullback_of_restructure` for a `Diagonal` to ensure that it returns a `Diagonal`. This is good -- it means we can occassionally get faster computations in the natural pullback. 
```
?
